core/src/main/java/org/infinispan/functional/impl/AbstractFunctionalMap.java

```diff
@@ -114,20 +101,71 @@ private Transaction getOngoingTransaction() {
       }
    }
 
-   protected Object invoke(InvocationContext ctx, VisitableCommand cmd) {
-      Object result;
+   protected <T> CompletableFuture<T> invokeAsync(InvocationContext ctx, VisitableCommand cmd) {
+      CompletableFuture<T> cf;
+      boolean isImplicitTx = ctx.isInTxScope() && ((TxInvocationContext) ctx).isImplicitTransaction();
+      final Transaction implicitTransaction;
       try {
-         result = fmap.chain().invoke(ctx, cmd);
+         cf = (CompletableFuture<T>) fmap.chain.invokeAsync(ctx, cmd);
+         if (isImplicitTx) {
+            implicitTransaction = transactionManager.suspend();
+         } else {
+            implicitTransaction = null;
+         }
+      } catch (SystemException e) {
+         throw new CacheException("Cannot suspend implicit transaction", e);
       } catch (Throwable t) {
-         try {
-            rollbackIfNeeded(ctx);
-         } catch (Throwable t2) {
-            t2.addSuppressed(t);
-            throw t2;
+         if (isImplicitTx) {
+            try {
+               if (transactionManager != null) transactionManager.rollback();
+            } catch (Throwable t2) {
+               log.trace("Could not rollback", t2);//best effort
+               t.addSuppressed(t2);
+            }
          }
          throw t;
       }
-      commitIfNeeded(ctx);
-      return result;
+      if (isImplicitTx) {
+         return cf.handle((result, throwable) -> {
+            if (throwable != null) {
+               try {
+                  implicitTransaction.rollback();
+               } catch (SystemException e) {
+                  log.trace("Could not rollback", e);
+                  throwable.addSuppressed(e);
+               }
+               throw CompletableFutures.asCompletionException(throwable);
+            }
+            try {
+               implicitTransaction.commit();
+            } catch (Exception e) {
+               log.couldNotCompleteInjectedTransaction(e);
+               throw CompletableFutures.asCompletionException(e);
+            }
+            return result;
+         });
+      } else {
+         return cf;
+      }
+   }
+
+   protected <T> CompletableFuture<T> joinIfNeeded(CompletableFuture<T> future) {
+      if (futureMode == Param.FutureMode.COMPLETED) {
+         joinForced(future);
+      }
+      return future;
+   }
+
+   // TODO: we shouldn't force joins, Traversables should behave lazily
```

Comment:
Looking at the `FutureMode` javadoc, it should only apply to methods that return `CompletableFuture`. But with these changes, perhaps `FutureMode.COMPLETED` isn't really needed any more? /cc @galderz